nntrainer/src/layers.cpp

```diff
@@ -182,65 +182,55 @@ float ReluPrime(float x) {
   }
 }
 
+static auto rng = [] {
+  std::mt19937 rng;
+  rng.seed(std::random_device()());
+  return rng;
+}();
+
+template <typename T, typename... Args>
+static T RandNormal(Args &&... args) {
+  std::normal_distribution<T> dist(std::forward<Args>(args)...);
+  return dist(rng);
+}
+
+template <typename T, typename... Args>
+static T RandUniform(Args &&... args) {
+  std::uniform_real_distribution<T> dist(std::forward<Args>(args)...);
+  return dist(rng);
+}
+
 static Tensors::Tensor WeightInitialization(unsigned int width, unsigned int height, Layers::weightIni_type init_type) {
-  std::random_device rd;
-  std::mt19937 gen(rd());
   Tensors::Tensor W = Tensors::Tensor(height, width);
 
-  if (init_type == Layers::WEIGHT_UNKNOWN)
-    init_type = Layers::WEIGHT_XAVIER_NORMAL;
-  switch (init_type) {
-    case Layers::WEIGHT_LECUN_NORMAL: {
-      std::normal_distribution<float> dist(0, sqrt(1 / height));
-      for (unsigned int i = 0; i < width; ++i)
-        for (unsigned int j = 0; j < height; ++j) {
-          float f = dist(gen);
-          W.setValue(0, j, i, f);
-        }
-    } break;
-    case Layers::WEIGHT_LECUN_UNIFORM: {
-      std::uniform_real_distribution<float> dist(-1.0 * sqrt(1.0 / height), sqrt(1.0 / height));
-      for (unsigned int i = 0; i < width; ++i)
-        for (unsigned int j = 0; j < height; ++j) {
-          float f = dist(gen);
-          W.setValue(0, j, i, f);
-        }
-    } break;
-    case Layers::WEIGHT_XAVIER_NORMAL: {
-      std::normal_distribution<float> dist(0, sqrt(2.0 / (width + height)));
-      for (unsigned int i = 0; i < width; ++i)
-        for (unsigned int j = 0; j < height; ++j) {
-          float f = dist(gen);
-          W.setValue(0, j, i, f);
-        }
-    } break;
-    case Layers::WEIGHT_XAVIER_UNIFORM: {
-      std::uniform_real_distribution<float> dist(-1.0 * sqrt(6.0 / (height + width)), sqrt(6.0 / (height + width)));
-      for (unsigned int i = 0; i < width; ++i)
-        for (unsigned int j = 0; j < height; ++j) {
-          float f = dist(gen);
-          W.setValue(0, j, i, f);
-        }
-    } break;
-    case Layers::WEIGHT_HE_NORMAL: {
-      std::normal_distribution<float> dist(0, sqrt(2.0 / (height)));
-      for (unsigned int i = 0; i < width; ++i)
-        for (unsigned int j = 0; j < height; ++j) {
-          float f = dist(gen);
-          W.setValue(0, j, i, f);
-        }
-    } break;
-    case Layers::WEIGHT_HE_UNIFORM: {
-      std::uniform_real_distribution<float> dist(-1.0 * sqrt(6.0 / (height)), sqrt(6.0 / (height)));
-      for (unsigned int i = 0; i < width; ++i)
-        for (unsigned int j = 0; j < height; ++j) {
-          float f = dist(gen);
-          W.setValue(0, j, i, f);
-        }
-    } break;
-    default:
-      W.setZero();
-      break;
+  for (unsigned int i = 0; i < width; ++i) {
+    for (unsigned int j = 0; j < height; ++j) {
+      float f;
+      switch (init_type) {
```

Comment:
Hmm.. I think the switch (or if) statement inside loop would be not good in the overall performance.
How about this style which adds common codes to the macro?
```
#define SOME_MACRO(x) for (...) for (...) ...

switch (init_type)
  case Layers::WEIGHT_LECUN_NORMAL:
     SOME_MARCO (RandNormal<float>(0, sqrt(1 / height)))
     break;
  ... 
```

Reply:
well. the code is not the newest. Thanks for the suggestion.